maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java

```diff
@@ -501,6 +504,12 @@
     @Parameter( property = "excludedGroups" )
     private String excludedGroups;
 
+    @Parameter( property = "junitIncludeEngine" )
```

Comment:
Both instance fileds are in wrong class. We have two plugins in this project, so you have to generate getter and setter for these fields. Both must be in `SurefirPlugin` class, see [this](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java#L99) and also in [here](https://github.com/apache/maven-surefire/blob/master/maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java#L116).
The reson is the prefix in the property. Due to AbstractSurefireMojo is an abstract class, you have to generate getters and setters and use them instead of fields in AbstractSurefireMojo.
Regarding the naming conventions pls start the name of fields and properties with "include" and "exclude", and raname junit to "junit5" to make it clear that we are aiming for not JUnit4 and 3.
Additionally both should be String[] because the history shows us that people want to have more than one value in our config properties, pls see [the tutorial for MOJO](https://maven.apache.org/guides/plugin/guide-java-plugin-development.html#parameter-types-with-multiple-values).

Reply:
I know it is quite a lot of work with the PR but it is always with a new feature. Pls add a Javadoc on the top of these fields and see how we use to write the Javadoc in another parts.
Additionally every feature has to be documented. Regarding JUnit5 add a new chanper with your feature [here](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/site/apt/examples/junit-platform.apt.vm).

Reply:
Hmm I follow simillar approach like for "groups" but now I am a little confused regarding those fields... 
Logic to load values for those fields is in AbstractSurefireMojo but if I will not define them here, so how I can get values for them?
Can you provide me some example field with simillar flow which you expect?

Reply:
yes, the trick is called an Abstcation.

Let's jump to the class `SurefirePlugin`, and consider the field [printSummary](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java#L100) and its [setter](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java#L702) and [getter](https://github.com/apache/maven-surefire/blob/master/maven-surefire-plugin/src/main/java/org/apache/maven/plugin/surefire/SurefirePlugin.java#L696).

Now jump to the class `IntegrationTestMojo`. The field with name [printSummary](https://github.com/apache/maven-surefire/blob/master/maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java#L117), [setter](https://github.com/apache/maven-surefire/blob/master/maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java#L692) and [getter](https://github.com/apache/maven-surefire/blob/master/maven-failsafe-plugin/src/main/java/org/apache/maven/plugin/failsafe/IntegrationTestMojo.java#L686).

These were concrete implementations. But notice that both classes implement the same interface through extension of the abstract class `AbstractSurefireMojo`, see [this](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/AbstractSurefireMojo.java#L162). So the [interface](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireExecutionParameters.java) declares abstract methods I spoke before, the [setter](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireExecutionParameters.java#L83) and [getter](https://github.com/apache/maven-surefire/blob/master/maven-surefire-common/src/main/java/org/apache/maven/plugin/surefire/SurefireExecutionParameters.java#L81).

The task for you is to rename your fields to `includeJUnit5Engines` and `excludeJUnit5Engines`. Then change the type to `String[]`. The IntelliJ IDEA will automatically update the getter and setter.
Write both abstract getter and setter in the interface `SurefireExecutionParameters`. Implement them in `SurefirePlugin` and `IntegrationTestMojo`. Move both fields [this](https://github.com/apache/maven-surefire/pull/344/commits/e8ef4f5153d5c3f82639e4ec8cd1b1364202dad8#diff-38e379eb63d7dcd2deb45902b4a494517e4ce54c29913bcf74e7d9243ebb7011R507) and [this](https://github.com/apache/maven-surefire/pull/344/commits/e8ef4f5153d5c3f82639e4ec8cd1b1364202dad8#diff-38e379eb63d7dcd2deb45902b4a494517e4ce54c29913bcf74e7d9243ebb7011R511) to `SurefirePlugin` and `IntegrationTestMojo`. Then remove the getters and setters in `AbstractSurefireMojo from Line 3815 to 3836. Try to compile the project.

Reply:
Ok, thank you. I will try to do this on Tuesday :)

Reply:
I expected to finish it today but ok I will wait for you till Tuesday.

Reply:
Hi, I made most of changes but there is conversion of String[] to String which does not look so good but I need this to pass values to properties.
Regarding that include and exclude engines should be provided in array I agree but please, remember that most of usage of this feature will base probably on use of one value only as junit as default have only one engine (one for JUnit5 and one for JUnit Vintage).
I will push my changes in next minutes